File: vsop87a_ephemeris.py
import numpy as np

import tools

class VSOP87AEphemeris:
    """
    Computes planet positions and velocities w.r.t. the Sun using VSOP87A series.
    """
    AU = 149597870.691  # AU in km

    def __init__(self, source):
        """
        Here source can be file path to json file containing the series or
        the json object itself.
        """
        if isinstance(source, str):
            source = tools.load_json(source)
        self.bodies = source['bodies']
        # Reshape the coefficients from (n,) to (n/3,3)
        for body in self.bodies.values():
            for group in body:
                group['coeffs'] = np.array(group['coeffs']).reshape(-1, 3)
        self.matrix = np.array(source['matrix'])

    def get_pos_vel(self, body_name: str, t: float):
        """
        Returns position and velocity for given body and time in Julian centuries since J2000.0.
        Units are km, Julian day.
        """
        # Convert from centuries to millenia:
        t = 0.1 * t
        t_pow = np.power(t, np.arange(6))
        t_pow_p = np.array([k * t_pow[k-1] if k > 0 else 0.0 for k in range(6)])

        pos = np.zeros(3)
        vel = np.zeros(3)
        for group in self.bodies[body_name]:
            coord = group['coord']
            alpha = group['alpha']
            coeffs = group['coeffs']

            sum_pos = np.sum(coeffs[:,0] * np.cos(coeffs[:,1] + t*coeffs[:,2]))
            sum_vel = -np.sum(coeffs[:,0] * coeffs[:,2] * np.sin(coeffs[:,1] + t*coeffs[:,2]))

            pos[coord] += t_pow[alpha] * sum_pos
            vel[coord] += t_pow_p[alpha] * sum_pos + t_pow[alpha] * sum_vel
        return self.matrix @ pos * self.AU, self.matrix @ vel * self.AU / 365250.0

if __name__ == '__main__':
    vsop87 = VSOP87AEphemeris('./json/vsop87a_raw.json')
    # vsop87 = VSOP87AEphemeris('./json/vsop87a_truncated_7.json')
    p, v = vsop87.get_pos_vel('EARTH-MOON', -5.25)
    print(p.tolist(), v.tolist())

File: javascript/vsop87aEphemeris.js
const { vectorMult, matrixMult } = require('./tools');

/**
 * Computes planet positions and velocities w.r.t. the Sun using VSOP87A series.
 */
class VSOP87AEphemeris {
    AU = 149597870.691;  // AU in km

    /**
     * Here source has to be object containing the (possibly truncated) VSOP87A series
     * loaded from one of the json files.
     */
    constructor(source) {
        this.bodies = source['bodies'];
        this.matrix = source['matrix'];
    }

    /**
     * Returns position and velocity for given body and time in Julian centuries since J2000.0. 
     * Units are km, Julian day.
     */
    getPosVel(bodyName, t) {
        const tPow = Array.from({ length: 6 }, (_, k) => Math.pow(t, k));
        const tPowP = Array.from({ length: 6 }, (_, k) => k > 0 ? k * Math.pow(t, k-1) : 0);
        
        let pos = [0, 0, 0];
        let vel = [0, 0, 0];
        this.bodies[bodyName].forEach((group) => {
            const coord = group['coord'];
            const alpha = group['alpha'];
            const coeffs = group['coeffs'];

            let sumPos = 0;
            let sumVel = 0;
            for (let k3 = 0; k3 < coeffs.length; k3 += 3) {
                const a = coeffs[k3 + 0];
                const b = coeffs[k3 + 1];
                const c = coeffs[k3 + 2];
                sumPos += a * Math.cos(b + c*t);
                sumVel -= a * c * Math.sin(b + c*t);
            }

            pos[coord] += tPow[alpha] * sumPos;
            vel[coord] += tPowP[alpha] * sumPos + tPow[alpha] * sumVel;
        });

        pos = vectorMult(matrixMult(this.matrix, pos), this.AU);
        vel = vectorMult(matrixMult(this.matrix, vel), this.AU / 36525);
        return { pos, vel };
    }
}

module.exports = VSOP87AEphemeris;